### PR TITLE
Rewrite func_log()

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This PR completely rewrites the func_log() function to be able to
parse `ws_value_*` types into their string representation and log them.

Pair-programmed-with: ~~Marcel Müller com2040@gmail.com~~ @TheNeikos 
Suggested-by: ~~Marcel Müller com2040@gmail.com~~ @TheNeikos

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Q: Shall I add the appropriate documentation in the user doc?

---

Note: I tested this transaction:

``` json
{
    "TYPE": "transaction",
    "UID": 1,
    "FLAGS": {
        "EXEC": true
    },
    "CMDS": [
        {"push": [true]},
        {"call": [{"pos":0}, "log", "[CONTEXT] ", 3, "FMT", {"pos": -1}]}
    ]
}
```

and it **works!!!**

**Comment by matthiasbeyer** [comment]
I just tested it with

``` json
{
    "TYPE": "transaction",
    "UID": 1,
    "FLAGS": {
        "EXEC": true
    },
    "CMDS": [
        {"push": [true]},
        {"push": [2]},
        {"push": [34]},
        {"push": [null]},
        {"push": ["ENDE"]},
        {"call": [{"pos":0}, "log", "[CONTEXT] ", 3, "FMT ", {"pos": -5}, " ", {"pos": -4}, " ", {"pos": -3}, " ", {"pos": -2}, " ", {"pos": -1}]}
    ]
}
```

**and it still works!!!** ... I think we're awesome! :smile:

**Comment by neithernut** [comment]
I think the conversion of values into strings should probably go into `ws_value_union` or something (would make things more reusable).

**Comment by matthiasbeyer** [comment]
Please review.

I created a to-string functionality for the `ws_value_union` type, which converts the value instances to strings like so:

``` plain
ws_value        -> "<value>"
ws_value_none   -> "<none>"
ws_value_nil    -> "<nil>"
ws_value_bool   -> "<bool:true>" | "<bool:false>"
ws_value_int    -> "<int:1>" (etc)
ws_value_string -> "<string:'blablubb'>"
```

I would like to add to-string functionality for all `ws_value_*` types, also for
the set, which would result in something like this:

``` plain
"<set:[...]>"
```

as I do not want to convert everything in a set to a string ... would be much to heavy!

Suggestions?

**Comment by neithernut** [comment]
I'd postpone this, as it would depend on functionality in the object.

File: src/util/string.h [hunk 1]
```diff
@@ -53,6 +53,16 @@
...
+/*
+ * Convert a hardcoded int to string
+ */
+
+#define STR_HELPER(x) #x
+
+/**
+ * int to string
+ */
+#define STR_OF(x) STR_HELPER(x)
```
Comment (neithernut) [inline src/util/string.h]: This is unnecessarily verbose.

File: src/context.c [hunk 2]
```diff
@@ -161,40 +165,55 @@ static int
...
 
-    return 1;
+    while (--n) {
+        if (fary[n]) {
+            free(ary[n]);
+        }
+    }
+
+    return 0;
 }
```
Comment (neithernut) [inline src/context.c:212]: There is no need for `fary`. Storing the index of the element you were processign when bailing out is sufficient for deinitialization in an error case.

File: src/util/string.h (hunk 1 above)
Comment (matthiasbeyer) [inline src/util/string.h]: What would you suggest?

Comment (neithernut) [inline src/util/string.h]: `#define STR_OF(x) #x`

Comment (matthiasbeyer) [inline src/util/string.h]: This does not work. You actually need a macro-in-macro construct for this!

Comment (neithernut) [inline src/util/string.h]: Ok.

Comment (TheNeikos) [inline src/util/string.h]: What about `##x` ?

**Comment by neithernut** [comment]
Is there any reason you do a `snprintf("%s%s%s")` to stringify values? Why not `<string:%s>` etc.?

File: src/util/string.h (hunk 1 above)
Comment (matthiasbeyer) [inline src/util/string.h]: I will try...

Comment (matthiasbeyer) [inline src/util/string.h]: nope.

**Comment by matthiasbeyer** [comment]
It was because of the length of the prefix and postfix, I didn't like to hard-code them. But I removed the prefix-postfix-foo anyways, as it was ugly to read in the logs. Hope you acknowledge that!

File: src/util/string.h (hunk 1 above)
Comment (TheNeikos) [inline src/util/string.h]: http://codepad.org/DOG1mhlo I am not sure why it's not doing what you want...?

Comment (matthiasbeyer) [inline src/util/string.h]: Hm, I had no parentheses... strange things are strange!

Comment (TheNeikos) [inline src/util/string.h]: :P Rule #1 in Macros, never _enough_ parantheses.

Comment (matthiasbeyer) [inline src/util/string.h]: This was no. 1 rule for lisp... didn't know it also applies to macros! :laughing:

**Comment by neithernut** [comment]
Mh... come to think of it, given a command like:

```
{"log" : ["Current surface id:", {pos: -1}]}
```

Do you really want this to end up like:

```
<string:Current surface id:><int:4>
```

?

**Comment by matthiasbeyer** [comment]
I actually stated in my previous comment that the code which would result in a log-message like this was _removed_ because I considered it as ugly! Now you have a simple concatenation of the log stuff, so your call would result in

```
Current surface id4
```

(I do not add whitespace)

**Comment by neithernut** [comment]
Yep, I just wanted to poitnt this out. Sorry for the confusion.
Anyways: I'm ok with this (ACK for squashing).

**Comment by matthiasbeyer** [comment]
Done.

**Comment by neithernut** [comment]
I think the alst three commits (https://github.com/matthiasbeyer/waysome/commit/36c17065744267d7921a403ab18ace785f201221, https://github.com/matthiasbeyer/waysome/commit/a9fb35dc3ca444be1216970abced2b4c6d57edba and https://github.com/matthiasbeyer/waysome/commit/b6db5ddadfb801fdbff463b08d438e9383286955) should be squashable, too.

**Comment by matthiasbeyer** [comment]
Squashed some more commits, not entirely as you suggested, but almost.

**Comment by neithernut** [comment]
ACK, but you have to rebase now...

**Pull request merged**